Microsoft.Toolkit.HighPerformance/Helpers/NullableHelper.cs

```diff
@@ -0,0 +1,35 @@
+using System;
+using System.Runtime.CompilerServices;
+
+
+namespace Microsoft.Toolkit.HighPerformance.Helpers
+{
+    /// <summary>
+    /// Helpers to perform operations on nullable value types
+    /// </summary>
+    public static class NullableHelper
+    {
+        /// <summary>
+        /// We use this type to publically expose the fields of a Nullable{T} type
+        /// </summary>
+        private struct NullableHelperMap<T>
+            where T : struct
+        {
+            public bool HasValue;
+            public T Value;
+        }
+
+        /// <summary>
+        /// Returns a ref <typeparamref name="T"/> to the value of <paramref name="nullable"/> if <see cref="Nullable{T}.HasValue"/> is true,
+        /// else, default(T)
+        /// </summary>
+        /// <typeparam name="T">The type of the underlying value</typeparam>
+        /// <param name="nullable">The <see cref="Nullable{T}"/></param>
+        /// <returns>A ref <typeparamref name="T"/> to the underlying value or default(T)</returns>
+        public static ref T GetValueOrDefaultRef<T>(this ref T? nullable)
+            where T : struct
+        {
+            return ref Unsafe.As<T?, NullableHelperMap<T>>(ref nullable).Value;
+        }
```

Comment:
My main issue with this PR is that I'm not convinced this API is designed correctly. If a user calls this and then writes a value to the returned reference, the `Nullable<T>` instance would be in an inconsistent state, as the `hasValue` field could potentially still be set to `false` despite the `value` field now being set to something other than `default`.

For correctness, this should actually be like:

```suggestion
        public static ref readonly T GetValueOrDefaultReference<T>(in this T? nullable)
            where T : struct
        {
            ref T? r0 = ref Unsafe.AsRef(nullable);
            ref RawNullableData<T> r1 = ref Unsafe.As<T?, RawNullableData<T>>(ref r0);
            
            return ref t1.Value;
        }
```